content/news/2019-05-28-alga-introduction.md

```diff
@@ -0,0 +1,120 @@
+author: 'Vasily Alferov'
+---
+
+I am a student at Summer of Haskell this year and I am very thankful to authors
+of this site who provided me this platform to have a blog about GSoC. This is
+an introduction post about my project. I am planning to write regular posts
+on my progress and when I find something interesting.
+
+The idea of the project was on the
+[ideas list](https://summer.haskell.org/ideas.html#algebraic-graphs) published
+earlier. There are two of us who were accepted for this project, the other one
+is Adithya Kumar and it is he who will be doing the part described on the list.
+He told me his GSoC blog will probably be
+[there](https://adithyaov.github.io/).
+
+My task is to introduce bipartite graphs to Alga and that is what I am going to
+tell about now.
+
+# Algebraic graphs
+
+There are three common ways to represent graphs in computing: adjacency matrix,
+adjacency lists, and edge lists. All three of them have their advantages and
+disadvantages. The most commonly used is the adjacency lists approach: that is
+storing a list of neighbors for each vertex. In fact, I can think of only one
+common algorithm for which this approach is not perfect: it is Cruscal's
+algorithm for finding the minimum spanning tree.
+
+The problem is that feeding graphs formed this way to algorithms is not safe.
+For example, if the algorithm is designed for bidirectional graphs, it may
+rely on the fact that if some vertex $u$ is in the list of neighbors of some
+another vertex $v$ then $v$ is in the list of neighbors of $u$.
+
+A traditional solution for functional programming would be to guarantee the
+consistency of input data for the algorithm by taking a representation of the
+graph that would not allow a wrong graph to be passed. That's what we call
+_type safety_.
+
+[Alga](https://github.com/snowleopard/alga) is a library that provides such
+representation with a beautiful algebraic interpretation. It also has a nice
+set of algorithms out of the box. You may read a paper on Alga by its author
+[here](https://github.com/snowleopard/alga-paper), I'm just going to provide
+some basics.
+
+Consider the following definition for the graph data type:
+
+```Haskell
+data Graph a = Empty
+             | Vertex a
+             | Overlay (Graph a) (Graph a)
+             | Connect (Graph a) (Graph a)
+```
+
+The constructors mean the following:
+
+* `Empty` constructs an empty graph.
+* `Vertex v` constructs a graph of single vertex labeled `v`.
+* `Overlay g h` constructs a graph with sets of unions and edges united from
+  graphs `g` and `h`.
+* `Connect g h` does the same as `Overlay` and also connects all vertices of
+  `g` to all vertices of `h`.
+
+One can easily construct a `Graph` of linear size having a list of edges of the
+desired graph. In fact, this approach may even save memory for dense graphs
+comparing to adjacency lists. And this approach is surely _type safe_ in the
+sense described above.
```

Comment:
Add some more detail: repeat that you can't connect to non-existing vertices here.